### PR TITLE
nvdimm_negative: Add case to test negative align option

### PR DESCRIPTION
**Pull request by huangyum**
Signed-off-by: Yumei Huang <yuhuang@redhat.com>
ID: 1903006

### CODE REVIEW

**Comment by huangyum** [comment]
(1/2) type_specific.io-github-autotest-qemu.nvdimm.nvdimm_negative.align_max: PASS (36.40 s)
 (2/2) type_specific.io-github-autotest-qemu.nvdimm.nvdimm_negative.align_illegal: PASS (11.57 s)
@MiriamDeng @yanan-fu Pls help review, thanks.

File: qemu/tests/nvdimm_negative.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    :param env: Dictionary with test environment.
+    """
+
+    vm = env.get_vm(params["main_vm"])
+    params['start_vm'] = 'yes'
+    error_msg = params.get('error_msg', '')
+    try:
+        vm.create(params=params)
+        output = vm.process.get_output()
+    except virt_vm.VMCreateError as e:
```
Comment (yanan-fu) [inline qemu/tests/nvdimm_negative.py:26]: Do we need to distinguish the different behaviour for different qemu version ? Create vm failed before QEMu 5.2.

Comment (huangyum) [reply qemu/tests/nvdimm_negative.py:26]: I don't think so. As long as no core dump and a corresponding message is prompted, I think it's fine.

File: qemu/tests/cfg/nvdimm.cfg [hunk 2]
```diff
@@ -92,3 +92,17 @@
...
+        - nvdimm_negative:
+            type = nvdimm_negative
+            start_vm = no
+            size_mem = 8G
+            pre_command = "dd if=/dev/zero of=${nv_backend} bs=1M count=8192"
+            variants:
+                - align_max:
+                    align_mem = 4G
+                    error_msg = "the alignment \(0x[0-9A-Fa-f]+\) is not supported|the alignment \(0x[0-9A-Fa-f]+\) exceeds the expected maximum alignment"
+                    ppc64le:
```
Comment (yanan-fu) [inline qemu/tests/cfg/nvdimm.cfg:104]: @MiriamDeng  @maxujun  do we need consisder `ppc64` ? Thanks!

Comment (yanan-fu) [reply qemu/tests/cfg/nvdimm.cfg:104]: Ignore me,  this case use `only x86_64, ppc64le`

File: qemu/tests/cfg/nvdimm.cfg [hunk 3]
```diff
@@ -92,3 +92,17 @@
...
             target_mems = mem1
+        - nvdimm_negative:
+            type = nvdimm_negative
+            start_vm = no
+            size_mem = 8G
+            pre_command = "dd if=/dev/zero of=${nv_backend} bs=1M count=8192"
+            variants:
+                - align_max:
+                    align_mem = 4G
+                    error_msg = "the alignment \(0x[0-9A-Fa-f]+\) is not supported|the alignment \(0x[0-9A-Fa-f]+\) exceeds the expected maximum alignment"
```
Comment (yanan-fu) [inline qemu/tests/cfg/nvdimm.cfg:103]: The align_mem `4G` already fixed, why don't use the `0x100000000` directly ? @huangyum

Comment (huangyum) [reply qemu/tests/cfg/nvdimm.cfg:103]: I'm not sure if the max value would change in the future, so use an exp.

Comment (yanan-fu) [reply qemu/tests/cfg/nvdimm.cfg:103]: > I'm not sure if the max value would change in the future, so use an exp.

It does not matter as my understanding.  If the max value be bigger than 4G in the feature, the `align_mem` still need be updated.  The value in the error_msg is match with the setting of `align_mem`.

Comment (huangyum) [reply qemu/tests/cfg/nvdimm.cfg]: Make sense. Updated. Thanks.

File: qemu/tests/cfg/nvdimm.cfg [hunk 4]
```diff
@@ -92,3 +92,17 @@
...
             target_mems = mem1
+        - nvdimm_negative:
+            type = nvdimm_negative
+            start_vm = no
+            size_mem = 8G
+            pre_command = "dd if=/dev/zero of=${nv_backend} bs=1M count=8192"
+            variants:
+                - align_max:
+                    align_mem = 4G
+                    error_msg = "the alignment \(0x100000000\) is not supported|the alignment \(0x100000000\) exceeds the expected maximum alignment"
```
Comment (yanan-fu) [inline qemu/tests/cfg/nvdimm.cfg:103]: `the alignment \(0x100000000\) (is not supported|exceeds the expected maximum alignment)` ?

Comment (MiriamDeng) [inline qemu/tests/cfg/nvdimm.cfg:103]: On ppc, "-device nvdimm" also needs "uuid" parameter.
And according the current parameters
size = 8G
lable_size=256M
align_mem =4G
The error output should be "backing store size 0x80000000 does not match 'size' option 0x200000000
" on ppc

File: qemu/tests/cfg/nvdimm.cfg [hunk 5]
```diff
@@ -92,3 +92,17 @@
...
+                    ppc64le:
+                        error_msg = ""
+                - align_illegal:
+                    align_mem = 1.5G
+                    error_msg = "alignment 0x60000000 must be a power of two"
```
Comment (MiriamDeng) [inline qemu/tests/cfg/nvdimm.cfg:108]: on ppc, error message should be "backing store size 0x80000000 does not match 'size' option 0x200000000"

File: qemu/tests/cfg/nvdimm.cfg [hunk 6]
```diff
@@ -92,3 +92,17 @@
...
+            type = nvdimm_negative
+            start_vm = no
+            size_mem = 8G
+            pre_command = "dd if=/dev/zero of=${nv_backend} bs=1M count=8192"
+            variants:
+                - align_max:
+                    align_mem = 4G
+                    error_msg = "the alignment \(0x100000000\) is not supported|the alignment \(0x100000000\) exceeds the expected maximum alignment"
+                    ppc64le:
+                        error_msg = ""
```
Comment (MiriamDeng) [inline qemu/tests/cfg/nvdimm.cfg:105]: And according the current parameters
size = 8G
lable_size=256M
align_mem =4G
The error output should be "backing store size 0x80000000 does not match 'size' option 0x200000000
" on ppc

File: qemu/tests/cfg/nvdimm.cfg (hunk 4 above)
Comment (huangyum) [reply qemu/tests/cfg/nvdimm.cfg:103]: Does every case need 'uuid' for ppc? What's the expected value for uuid?
Besides, pls update the error message to polarion,  it says no error for ppc.

File: qemu/tests/cfg/nvdimm.cfg (hunk 5 above)
Comment (huangyum) [reply qemu/tests/cfg/nvdimm.cfg:108]: Again, please update steps in polarion first. If the error message is different with different versions of qemu, pls make it clear too.

File: qemu/tests/cfg/nvdimm.cfg (hunk 4 above)
Comment (MiriamDeng) [reply qemu/tests/cfg/nvdimm.cfg:103]: On ppc, the combination paramters are 2G,2G and 128K in the test case but now is 8G, 4G and 256M. but I can accept them, please get uuid by uuidgen,and polarion is done as well.

File: qemu/tests/cfg/nvdimm.cfg (hunk 5 above)
Comment (MiriamDeng) [reply qemu/tests/cfg/nvdimm.cfg:108]: size = 2G, align = 1.5G and lable_size = 128K, please see test case in the ploarion.

Comment (huangyum) [reply qemu/tests/cfg/nvdimm.cfg:109]: The test point is align, as long as align value is as expected, other options would not affect.

**Comment by MiriamDeng** [comment]
JOB LOG    : /root/avocado/job-results/job-2020-12-03T04.52-56a6efd/job.log
 (1/2) type_specific.io-github-autotest-qemu.nvdimm.nvdimm_negative.align_max: PASS (33.35 s)
 (2/2) type_specific.io-github-autotest-qemu.nvdimm.nvdimm_negative.align_illegal: PASS (6.09 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2020-12-03T04.52-56a6efd/results.html

**Comment by MiriamDeng** [comment]
LGTM

**Review by yanan-fu — APPROVED** [review]
Ack
